README.md
# OpenAI - Developer

This VSCode Extension queries the OpenAI models [OpenAI](https://openai.com/blog). The models currently implemented are ChatGPT and Codex. Other OpenAI models will be coming soon.

Each model has different capabilities and price points. Prices are per 1,000 tokens. You can think of tokens as pieces of words, where 1,000 tokens are about 750 words. See [OpenAI Pricing](https://openai.com/pricing)

> Missing something? Let us know by opening an issue on the [GitHub repository](https://github.com/mrsahugit/openai-developer/issues/new/choose)

## Features

<img src="https://raw.githubusercontent.com/mrsahugit/openai-developer/main/resources/media/help.gif">

### Start Conversation

Use the "OpenAI Developer: Start Conversation" command in the Command Palette:

1. Press `Ctrl` + `Shift` + `P` to open the Command Palette.
2. Begin typing `OpenAI Developer: Start Conversation`, and select it when it appears.
3. Enter the question you'd like to ask.
4. Wait for the response. It will appear in a new tab.

### Explain the code

<img src="https://raw.githubusercontent.com/mrsahugit/openai-developer/main/resources/media/explain-code.png">

Use the "OpenAI Developer: Explain the code" command in the Command Palette:

1. Select the block of code in the editor you need an explanation for.
2. Right click inside the editor and select `OpenAI Developer: Explain the code`.
4. Wait for the response. It will appear in a new tab.

### Why code is not working

Use the "OpenAI Developer: Why code is not working" command in the Command Palette:

1. Select the block of code in the editor you need to find out why not working
2. Right click inside the editor and select `OpenAI Developer: Why code is not working`.
4. Wait for the response. It will appear in a new tab.

### Change API Key

Once your OpenAI API key has been set you can update it using the "OpenAI Developer: Change API Key" command in the Command Palette:

1. Press `Ctrl` + `Shift` + `P` to open the Command Palette.
2. Begin typing `OpenAI Developer: Change API Key`, and select it when it appears.
3. Enter your OpenAI API key (see above instructions on how to find).

### Generate Image

<img src="https://raw.githubusercontent.com/mrsahugit/openai-developer/main/resources/media/generate-image.png">

To generate desired image(s) from a text description

1. Press `Ctrl` + `Shift` + `P` to open the Command Palette.
2. Begin typing `OpenAI Developer: Generate Image`, and select it when it appears.
3. Provide a description of the image.
4. Wait for the response. It will appear in a new tab.

## Authentication

When you query for the first time, you will be prompted to enter an OpenAI API key. This is used by the extension to access the API and is only sent to OpenAI.

### Obtaining API key

To find your OpenAI API key:

1. Go to https://platform.openai.com/account/api-keys. You will need to log in (or sign up) to your OpenAI account.
2. Click "Create new secret key", and copy it.
3. You should then paste it into VS Code when prompted.

## Configuration

### Switching Between Models

In order to switch between the models ChatGPT and Codex, follow these instructions:

1. `Ctrl` + `,` to open Preferences.
2. Search for openai-developer.model.
3. Use the dropdown to switch between ChatGPT and Codex.

<img src="https://raw.githubusercontent.com/mrsahugit/openai-developer/main/resources/media/settings.png">

### Maximum Tokens

To set the maximum number of tokens for each request follow these instructions:
> The maximum number of tokens to generate in the chat completion. The total length of input tokens and generated tokens is limited by the model's context length. 1,000 tokens are about 750 words.

1. `Ctrl` + `,` to open Preferences.
2. Search for openai-developer.maxTokens.
3. Use a number between 0 to Infinity. Default is set to 1024.

### Temperature

To set the temperature follow these instructions:
> What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.

1. `Ctrl` + `,` to open Preferences.
2. Search for openai-developer.temperature.
3. Use a number between 0 to 2. Default is set to 0.5.

## Commands

| Command                                               |                                                                |
| ----------------------------------------------------- | -------------------------------------------------------------- |
| `OpenAI Developer: Change API Key`                    | Change the Open API Key to set Open AI account detail.         |
| `OpenAI Developer: Start Conversation`                | Start the conversation with OpenAI. Ask anything...            |
| `OpenAI Developer: Explain the code`                  | Select your code and find the explanation for your code.       |
| `OpenAI Developer: Why code is not working`           | Select your code and find out why the code is not working.     |
| `OpenAI Developer: Generate Image`                    | Generate desired image(s) from a text description.             |

## Requirements

You must have Visual Studio Code 1.75.0 (January 2023) or later to install the extension.

## Known Issues

## Release Notes

### 0.0.4
- `OpenAI Developer: Generate Image` command added

### 0.0.3

- MaxTokens and Temperature configurations added
- `OpenAI Developer: Explain the code` command added
- `OpenAI Developer: Why code is not working` command added

### 0.0.2

- Documentation Bug fixes

### 0.0.1

- Initial release of OpenAI Developer
- `OpenAI Developer: Start Conversation` command added

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "openai-developer" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/views/ChatProvider.ts
/**
 * @author Manas Sahu
 * https://github.com/mrsahugit/openai-developer
 *
 * @license
 * Copyright (c) 2023 - Present, Manas Sahu
 *
 * All rights reserved. Code licensed under the MIT license
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 */
import * as vscode from 'vscode';

export class ChatProvider implements vscode.WebviewViewProvider {

    private _view?: vscode.WebviewView;

    resolveWebviewView(webviewView: vscode.WebviewView, context: vscode.WebviewViewResolveContext<unknown>, token: vscode.CancellationToken): void | Thenable<void> {
        this._view = webviewView;

        webviewView.webview.options = {
            enableScripts: true,
        };

        webviewView.webview.html = this.getWebviewContent();

        webviewView.webview.onDidReceiveMessage(
            message => {
                switch (message.command) {
                    case 'query':
                        vscode.window.showErrorMessage(message.text);
                        return;
                }
            },
            undefined,
        );
    }

    private getWebviewContent() {
        return `<!doctype html>
                <html lang="en">
                <head>
                    <meta charset="utf-8">
                    <meta name="viewport" content="width=device-width, initial-scale=1">
                    <title>OpenAI Developer</title>
                    <style>
                    .container {
                        padding: 0.2rem 0.2rem;
                      }
                      
                      .txt {
                        border: 1px solid #CCC;
                        border-radius: 0.375rem;
                        padding: 0.75rem 0.6rem;
                      }
                      .txt:focus {
                        outline: none;
                        border: 1px solid #2C94F1;
                      }
                      
                      .w100 {
                        width: 100%;
                      }
                      .tacenter {
                        text-align: center;
                      }
                      
                      .btn {
                        border: 1px solid #CCC;
                        border-radius: 0.375rem;
                        padding: 0.75rem 1.25rem;
                        float: right;
                      }
                      
                      .lblLarge {
                        font-size: 1.2rem;
                      }
                      
                      .lblMedium {
                        font-size: 1.0rem;
                        margin-top: 0.5rem;
                        margin-bottom: 0.25rem;
                      }
                      
                      .lblSmall {
                        font-size: 0.9rem;
                      }
                      
                      .lblMini {
                        opacity: 0.75;
                        font-size: 0.75rem;
                      }
                    </style>
                </head>
                <body>
                    <div class="container">
                        <input type="button" value="Send" class="btn" id="send" onclick="search();" />
                        <div style="overflow: hidden; padding-right: 2.0rem;">
                            <input type="text" class="txt w100" id="q" placeholder="Ask anything..." />
                        </div>
                    </div>
                    <div class="container">
                        <div class="lblLarge">Advanced Options</div>
                        <div>
                            <div class="lblMedium">Please select your model</div>
                            <input type="radio" id="chatgpt" name="models" value="chatgpt" checked />
                            <label for="chatgpt" class="lblSmall">ChatGPT</label><br>
                            <input type="radio" id="codex" name="models" value="codex">
                            <label for="codex" class="lblSmall">Codex</label><br>  
                        </div>
                        <div>
                            <div class="lblMedium">Temperature</div>
                            <div class="lblMini">What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.</div>
                            <input type="number" class="txt tacenter" id="temperature" min="0.0" max="2.0" value="0.5" step="0.1" size="3" />
                        </div>
                        <div>
                            <div class="lblMedium">Max Tokens</div>
                            <div class="lblMini">The maximum number of tokens to generate in the chat completion. The total length of input tokens and generated tokens is limited by the model's context length.</div>
                            <input type="number" class="txt tacenter" id="max_tokens" min="100" value="1024" step="100" size="3" />
                        </div>
                    </div>
                    <script>
                    function search() {
                        let temperature = document.getElementById("temperature");
                      alert(temperature.value);
                    }
                    </script>
                </body>
                </html>`;
    }
}